### ISSUE TITLE
build: the artifacts of each build should be made available

### ISSUE THREAD

**Issue by bassam** [rook/rook#435]
we would like every build from master to be "published" so that it can be tested. Once all testing is complete we can decide to "promote" the build to a release (and not have to rebuild when we do so).

**Comment by bassam** [comment]
Here's a rough design for this.

We currently have the following release artifacts:

- binaries (rook, rookd, rook-operator)
- symbols (rook.debug, rookd.debug, rook-operator.debug)
- containers (rookd, rookd-operator, toolbox)

The `release` target should continue to generate all release artifacts, and the `publish` target should push all artifacts to the appropriate destinations.

Everytime we build the version number will be set to the output of `git describe --always --tags`. So if this is a build from master right after 0.3 it would look something like `v0.3.0-2-g770ebbc`.

If the version is built from a tagged release like `v0.4` then the `publish` target will promote the release to the appropriate channels. So our release flow becomes as follows:

1. changes happen as part of PRs
2. PRs are merged into master after review
3. Jenkins builds every commit from master and publishes it with a "dirty" version `v0.3.0-2-g770ebbc`
4. the artifacts are available for this build and can be tested
5. After all testing is complete and we approve a release we can tag a specific commit
6. Jenkins builds the new tag and publishes it and promotes it

**Comment by jbw976** [comment]
Note that the toolbox container is not currently part of the automated build scripts: #411 

Would *all* the artifacts be published after every commit to master? Obviously the containers would need to be published so that CI/CD could make use of them, but the binaries and debug symbols (which currently get published to github) i'm not so sure about.

**Comment by bassam** [comment]
yes all artifacts get published to S3 and quay.io. Github releases only show up when we "promote" the release.

**Comment by DanKerns** [comment]
How do you get the build id from a core file?
Why not just store the .debug files in the same directory structure as releases?
Are the debug files for a build in the rook-v0.3.0-linux-amd64-debug.tar.gz file and if so so you need the /symbols hierarchy?

**Comment by bassam** [comment]
@DanKerns I think it might be simpler to do away with the /symbol hierarchy for now as you suggest.

**Comment by travisn** [comment]
What about publishing builds during the development cycle? For example, if I want to publish a private build that can be shared with others on the team. Seems like I should publish to my own repository such as quay.io/travisn/rook-operator

**Comment by bassam** [comment]
@travisn I think you should use your own repo for that.

**Comment by bassam** [comment]
work on this is tracked in #505

**Comment by bassam** [comment]
closed via #505
